### DISCUSSION TITLE
Trying to understand cursor style

### DISCUSSION THREAD

**Discussion by nickzelei** [ghostty-org/ghostty#3836] [Q&A] [answered]
Hi there,

Trying to make sense of the cursor style. I am using `zsh` and have a pretty bare [setup](https://github.com/nickzelei/zshrc-config) on an M1 Max MBP Sequoia 15.1.1.

Coming from iTerm2, my cursor was set by default (I believe?) to unblinking box cursor.
I noticed in Ghostty that my cursor is set to blinking vertical bar.

I set out to change that as I prefer the solid box.

I tried setting the following settings:

```
cursor-style = box
cursor-style-blink = false
```

Reloaded ghostty, reloaded config, however these options seem to be being completely ignored.

Based on what I've read about the shell integration, I found the following settings ultimately get me what I want, but it's unclear to me as to _why_ this accomplishes this:

```
cursor-style-blink = false
shell-integration-features = no-cursor
```

What about turning off the cursor shell integration feature sets the default to `block` (maybe this is the ghostty default if there is no shell-integration?)

I think as a user I would have expected that regardless of the shell integration, setting the cursor style in the settings would have overridden this as otherwise I'm left feeling like I did something wrong or the settings just don't work.

I got AI to tell me to drop this in my `.zshrc` config: `echo -ne '\e[1 q` - which I got to mostly work when fiddling with various options, but I ended up disabling that in favor of the pure ghostty config.

I'd like to ideally go back to a 0 config ghostty setup if what I need to do is configure my `.zshrc` config correctly, but then again that is also shell specific, so having the editor do it for me is kind of nice too.

As my closing thoughts, I'm not super versed in shell configuration, so there may be some assumptions being made here about shell config that I'm not used to (as pretty much only ever being in iterm2 user on mac.)

**Comment by 00-kat** [comment, marked as answer]
The shell integration sets the cursor to a blinking bar in vi insert mode, a box in normal mode, etc.

Disabling the cursor part of the shell integration disables this and only this.

So yes, the configuration you have is indeed what you want:
```ini
cursor-style = block
cursor-style-blink = false
shell-integration-features = no-cursor
```

**Reply by thiagowfx** [reply]
s/box/block

**Reply by nickzelei** [reply]
> The shell integration sets the cursor to a blinking bar in vi insert mode, a box in normal mode, etc.
> 
> Disabling the cursor part of the shell integration disables this and only this.
> 
> So yes, the configuration you have is indeed what you want:
> 
> ```ini
> cursor-style = box
> cursor-style-blink = false
> shell-integration-features = no-cursor
> ```

The `cursor-style` only works though when `shell-integration-features = no-cursor` is present. 
I'm not using vi or anything (unless I'm midunderstanding what you mean here, but I just open the terminal and am testing out how the cursor looks).

Also changing the `shell-integration-features` seems to require a full restart of ghostty instead of just reloading the configuration (maybe this is intended). But with the `no-cursor` disabled, I can freely change the `cursor-style` and with a config reload it updates as expected.

**Reply by 00-kat** [reply]
> seems to require a full restart of ghostty

It injects code into your shell.  You can't uninject code without restarting ghostty or restarting your shell then sourcing the shell integration code again.

> I'm not using vi or anything (unless I'm midunderstanding what you mean here)

You're not; I'm just saying *why* it does that.

> s/box/block

Oops, fixed.

**Reply by b11e** [reply]
> > The shell integration sets the cursor to a blinking bar in vi insert mode, a box in normal mode, etc.
> > Disabling the cursor part of the shell integration disables this and only this.
> > So yes, the configuration you have is indeed what you want:
> > ```ini
> > cursor-style = box
> > cursor-style-blink = false
> > shell-integration-features = no-cursor
> > ```
> 
> The `cursor-style` only works though when `shell-integration-features = no-cursor` is present. I'm not using vi or anything (unless I'm midunderstanding what you mean here, but I just open the terminal and am testing out how the cursor looks).
> 
> Also changing the `shell-integration-features` seems to require a full restart of ghostty instead of just reloading the configuration (maybe this is intended). But with the `no-cursor` disabled, I can freely change the `cursor-style` and with a config reload it updates as expected.

EDIT: i forgot to kill tmux-server and then restart 🤦‍♂️ 

Does this stick?
I applied this and it only worked temporarily. Starts to blink again randomly.
Now I reload the config and restart ghostty and it blinks even though I have these in my config.

**Reply by Elmer-Almeida** [reply]
thank you!

**Reply by nyngwang** [reply]
> The shell integration sets the cursor to a blinking bar in vi insert mode, a box in normal mode, etc.
> 
> Disabling the cursor part of the shell integration disables this and only this.

@00-kat I'm here just to make sure that the "vi insert mode" here is not related to the insert mode in neovim, right? More specifically, does setting `no-cursor` to `shell-integration-features` has any effect on neovim? Ty for reading.

**Reply by 00-kat** [reply]
Yep, it's unrelated to Neovim.  It's a feature that only affects your shell.

**Comment by diego898** [comment]
Just to make sure I understand - you cannot have a block cursor if shell integration is present? (on macos)

**Reply by 00-kat** [reply]
You can—you just need to disable that specific shell-integration feature as follows:
```ini
shell-integration-features = no-cursor
```

**Reply by diego898** [reply]
Oh I'm sorry I misunderstood - thank you!

**Comment by magnusriga** [comment]
@00-kat Apologies for beating a dead horse here, but is there any way to maintain the ZSH vi-mode behavior of a line-style cursor in Insert mode, and a box-style cursor in Normal mode, while also removing the blinking?

Setting `shell-integration-features = no-cursor` leaves me with a box-style cursor in both Insert and Normal mode. With this setup, I can turn off the blinking with `cursor-style-blink = false`, thus solving the original problem, but introducing a larger one.

Setting `shell-integration-features = cursor` will give me a line-style cursor in Insert mode, and a box-style cursor in Normal mode, as expected, but with this setup I am unable to turn off the blinking. `cursor-style-blink = false` has no effect, as pointed out by other commenters.

Any and all help appreciated.

**Reply by 00-kat** [reply]
I only know how to do that for Bash and Fish, sorry.  I believe you did find https://github.com/ghostty-org/ghostty/discussions/2812#discussioncomment-12149663 to work though.

**Reply by magnusriga** [reply]
@00-kat Yep, thank you. Ended up re-writing it as seen in the last post, works well now. Had to manually set shell integrations variables, e.g. `export GHOSTTY_SHELL_INTEGRATION_NO_CURSOR=1`, over ssh, but that might be fixed later if the team decides to make a new ssh command with automatic shell integration.

**Comment by slavanorm** [comment]
this question makes the vi integration cursor defaults questionable, especially when they cannot be configured.
